### PR TITLE
fix: report full tree properly

### LINKED ISSUE #15: Only one child of a process is reported for systems with ps
If some process in the chain has more than one child, all of them except one are not reported.

**Environment:** Ubuntu 18.04

### Steps to reproduce
1.	Run the following command from a shell:
```bash
bash -c "sleep 1000 & sleep 1000" &
```
2.	Notice the pid of the `bash` process
3.	Run the following script:
```javascript
const pstree = require('pstree.remy');

pstree(<noted pid>, (error, children) => {
    console.log(children);
})
```

### Expected result
There is an array with two child pids printed to the console

### Actual result
The array contains a single pid.

### PR DESCRIPTION
**Pull request by axxie**
Fixes #15

### CODE REVIEW

**Comment by axxie** [comment]
@remy, please take a look.

**Comment by remy** [comment]
Waiting on changes as per comments on the code.

**Comment by axxie** [comment]
> Waiting on changes as per comments on the code.

@remy, what comments on the code?
Could you, please, point out where are those comments?

**Review by remy — CHANGES_REQUESTED** [review]

**Comment by remy** [comment]
For heaven's sake. Githubs UI for code reviews really isn't obvious. I posted the comment over a week ago, and returning now I see a tiny "pending" label. I've submitted the comments so you should be able to see them now (sorry about that)

File: tests/index.test.js [hunk 1]
```diff
@@ -33,7 +33,7 @@ test('can read full child process tree', t => {
 
       console.log(children);
 
-      t.equal(children.length, 2);
+      t.equal(children.length, 4);
```
[the remark this answers is not in the document]
Comment (axxie) [reply tests/index.test.js]: Done.

**Comment by axxie** [comment]
@remy, I've made the requested changes, please take a look.

**Comment by remy** [comment]
Sorry it's taken so long to answer this. But the pstree.remy module is designed to _only_ get the sub-tree and not the main process. This is how it's used in nodemon too. This current PR thus adds an additional pid that isn't wanted in the list.

Sorry for wasting your time.

**Comment by axxie** [comment]
> Sorry it's taken so long to answer this. But the pstree.remy module is designed to _only_ get the sub-tree and not the main process. This is how it's used in nodemon too. This current PR thus adds an additional pid that isn't wanted in the list.

@remy, I don't get it. This change doesn't get the pid of the main process. It only gets the sub-tree.

**Comment by axxie** [comment]
> @remy, I don't get it. This change doesn't get the pid of the main process. It only gets the sub-tree.

Actually, cross it. I see it.
But the fix is trivial.
And the purpose of the change is to get all children, in contrast to what current implementation does.
Do you mind if I make a fix you to not return the main process?

**Comment by axxie** [comment]
Apparently, there is some misunderstanding.
The purpose of this PR is to implement getting all children in cases, when some process in the chain has more than one child. Current implementation in the `master` doesn't do it, because it only remembers last parent. The proposed fix remembers all potential parents.

Returning the pid of the main process is unintentional side effect and it can be easily fixed.

**Comment by remy** [comment]
Just trying to get it straight in my head (and I'm very careful about these low level changes because it has 100% affect on Mac and *nix users of nodemon).

The original test case you raised in #15 - this creates 2 sub-processes (off of `bash`). I'd expect pstree to return 2 pids.

I _think_ you're talking about another test case where you have a sub-sub-process? Is that right?

**Comment by i-keliukh** [comment]
Actually, it is the same test case: it doesn't matter what process is parent of two children: the main process or some intermediate process in the chain.
In fact, originally I reproduced it with npm-run-all, not bash. As far as I remember, if  some npm task containt npm-run-all, the tree is like this:
```
nodemon
 +- npm
      +- npm-run-all
           +- npm
           |    +- node
           +- npm
                +- node
```

The issue with the current implementation is that once it went along one branch, it cannot return up to switch to another branch. This is because it only remembers one parent.

File: lib/tree.js [hunk 2]
```diff
@@ -17,10 +17,10 @@ module.exports = function(rootPid, callback) {
         .slice(1)
         .map(_ => _.trim())
         .reduce((acc, line) => {
-          if (line.indexOf(rootPid + ' ') === 0) {
-            const pid = line.split(/\s+/).pop();
-            acc.push(parseInt(pid, 10));
-            rootPid = pid;
```
Comment (remy) [inline lib/tree.js:23]: For future me, it was _this_ line (23) that caused the original problem of losing track of split trees.

**Review by remy — APPROVED** [review]

**Comment by remy** [comment]
I can't release this at the moment, local tests are failing (as I was worried they would - it's reporting half the number of processes found). 

I'm investigating atm.

**Comment by remy** [comment]
Live. What a fucking nightmare: https://github.com/remy/til/blob/master/node/spawn-sh-on-mac.md

**Pull request merged**